README.md
# Amazon Products Plugin

The **Amazon Products** Plugin is for [Grav CMS](http://github.com/getgrav/grav). Displays Amazon products by ASIN/ISBN-10 on your pages.

## Installation

Installing the Amazon plugin can be done in one of two ways. The GPM (Grav Package Manager) installation method enables you to quickly and easily install the plugin with a simple terminal command, while the manual method enables you to do so via a zip file.

### GPM Installation (Preferred)

The simplest way to install this plugin is via the [Grav Package Manager (GPM)](http://learn.getgrav.org/advanced/grav-gpm) through your system's terminal (also called the command line).  From the root of your Grav install type:

    bin/gpm install amazon-products

This will install the Amazon plugin into your `/user/plugins` directory within Grav. Its files can be found under `/your/site/grav/user/plugins/amazon-products`.

### Manual Installation

> NOTE: This plugin depends on [Shortcode Core](https://github.com/getgrav/grav-plugin-shortcode-core), so you need to install it in advance.

To install this plugin, just download the zip version of this repository and unzip it under `/your/site/grav/user/plugins`. Then, rename the folder to `amazon-products`. You can find these files on [GitHub](https://github.com/kinformation/grav-plugin-amazon-products) or via [GetGrav.org](http://getgrav.org/downloads/plugins#extras).

You should now have all the plugin files under

    /your/site/grav/user/plugins/amazon-products

> NOTE: This plugin is a modular component for Grav which requires [Grav](http://github.com/getgrav/grav) and the [Error](https://github.com/getgrav/grav-plugin-error) and [Problems](https://github.com/getgrav/grav-plugin-problems) to operate.

## Configuration

Before configuring this plugin, you should copy the `user/plugins/amazon-products/amazon-products.yaml` to `user/config/plugins/amazon-products.yaml` and only edit that copy.

Here is the default configuration and an explanation of available options:

```yaml
enabled: true
keys:
  accessKeyId: "your AWS Access Key ID"
  secretAccessKey: "your AWS Secret Access Key"
  associateTag: "your Amazon AssociateTag"
locale: "US"
```

* accessKeyId: (required) AWS Access Key ID for Product Advertising API.
* secretAccessKey: (required) AWS Secret Access Key for Product Advertising API.
* associateTag: Amazon Associate Tag.
* locale: (Default:US) Amazon locale. Select US, CA, CN, DE, ES, FR, IT, JP or UK.

This plugin requires [Amazon Product Advertising API](https://affiliate-program.amazon.com/gp/advertising/api/detail/main.html) Access Key ID and Secret Access Key.
Please register with AWS and obtain Access Key ID and Secret Key.

## Usage

```
[amazon asin=059035342X][/amazon]
or
[amazon asin=https://www.amazon.com/Harry-Potter-Sorcerers-Stone-Rowling/dp/059035342X/ref=sr_1_1...][/amazon]
```
* asin: (required) ASIN (ProductID of Amazon) or ISBN-10 (Book) or Amazon Products URL

### Book

```
[amazon asin=059035342X][/amazon]
```
![](assets/image/sample_book.png)

### Music CD

```
[amazon asin=B01LX4X7K0][/amazon]
```
![](assets/image/sample_music.png)

### Video

```
[amazon asin=B01B80CM4W][/amazon]
```
![](assets/image/sample_dvd.png)

### Other

```
[amazon asin=B015WXL0C6][/amazon]
```
![](assets/image/sample_other.png)

> NOTE: If you do not see your ad, please check the following.
> * The validity of your AWS Access Key IDs.
> * Whether you can acquire the item XML you specified.
You can check with [Product Advertising API Scratchpad](http://webservices.amazon.com/scratchpad/)
> * Check config of PHP extension `suhosin`.
`suhosin.executor.include.whitelist = phar, plugin`

CHANGELOG.md
# v0.5.0
##  12/10/2017

1. [](#improved)
   * fix: Anonymize form of AWS Secret Access Key

1. [](#new)
    * add: Editor button

# v0.4.0
##  09/28/2017

1. [](#improved)
   * add: Extract ASIN by raw amazonURL

# v0.3.0
##  08/08/2017

1. [](#improved)
   * update: Improve ad template
   * update: Optimize API request generation method
   * update: Optimize distribution of ProductGroup

# v0.2.1
##  08/04/2017

1. [](#improved)
    * update: Modify docs url

# v0.2.0
##  07/30/2017

1. [](#improved)
    * update: Plugin name change (amazon -> amazon-products)

# v0.1.0
##  07/26/2017

1. [](#new)
    * First Release!


amazon.php
<?php
/**
 * Short description for amazon.php
 *
 * @package amazon
 * @author Kazuya Kanatani
 * @version 0.1
 * @copyright (C) 2017 kinformation<kanatani.social@gmail.com>
 * @license MIT
 */

namespace Grav\Plugin;

use Grav\Common\Plugin;

class AmazonPlugin extends Plugin
{
    /**
     * @return array
     */
    public static function getSubscribedEvents()
    {
        return [
            'onShortcodeHandlers' => ['onShortcodeHandlers', 0],
            'onTwigTemplatePaths' => ['onTwigTemplatePaths', 0],
            'onTwigSiteVariables' => ['onTwigSiteVariables', 0],
        ];
    }

    /**
     * Add style and script to page.
     */
    public function onTwigSiteVariables()
    {
        $this->grav['assets']->addCss('plugin://amazon/assets/css/amazon.css');
    }

    /**
     * Add current directory to twig lookup paths.
     */
    public function onTwigTemplatePaths()
    {
        $this->grav['twig']->twig_paths[] = __DIR__ . '/templates';
    }

    /**
     * Initialize configuration
     */
    public function onShortcodeHandlers()
    {
        $this->grav['shortcode']->registerAllShortcodes(__DIR__.'/shortcodes');
    }
}


amazon_request.php
<?php
/**
 * Short description for amazon_request.php
 *
 * @package amazon-products
 * @author Kazuya Kanatani
 * @version 0.2
 * @copyright (C) 2017 kinformation<kanatani.social@gmail.com>
 * @license MIT
 */

$locales = array(
    'US' => array(
        'domain'             => 'Amazon.com',
        'baseUri'            => 'https://webservices.amazon.com/onca/xml',
        'linkTemplate'       => '<iframe src="https://rcm.amazon.com/e/cm?t=${t}&o=1&p=8&l=as1&asins=${asins}&fc1=${fc1}&IS2=${IS2}&lt1=${lt1}&m=amazon&lc1=${lc1}&bc1=${bc1}&bg1=${bg1}&f=ifr" style="width:120px;height:240px;" scrolling="no" marginwidth="0" marginheight="0" frameborder="0"></iframe>',
    ),
    'UK' => array(
        'domain'             => 'Amazon.co.uk',
        'baseUri'            => 'https://webservices.amazon.co.uk/onca/xml',
        'linkTemplate'       => '<iframe src="https://rcm-uk.amazon.co.uk/e/cm?t=${t}&o=2&p=8&l=as1&asins=${asins}&fc1=${fc1}&IS2=${IS2}&lt1=${lt1}&m=amazon&lc1=${lc1}&bc1=${bc1}&bg1=${bg1}&f=ifr" style="width:120px;height:240px;" scrolling="no" marginwidth="0" marginheight="0" frameborder="0"></iframe>',
    ),
    'DE' => array(
        'domain'             => 'Amazon.de',
        'baseUri'            => 'https://webservices.amazon.de/onca/xml',
        'linkTemplate'       => '<iframe src="https://rcm-de.amazon.de/e/cm?t=${t}&o=3&p=8&l=as1&asins=${asins}&fc1=${fc1}&IS2=${IS2}&lt1=${lt1}&m=amazon&lc1=${lc1}&bc1=${bc1}&bg1=${bg1}&f=ifr" style="width:120px;height:240px;" scrolling="no" marginwidth="0" marginheight="0" frameborder="0"></iframe>',
    ),
    'FR' => array(
        'domain'             => 'Amazon.fr',
        'baseUri'            => 'https://webservices.amazon.fr/onca/xml',
        'linkTemplate'       => '<iframe src="https://rcm-fr.amazon.fr/e/cm?t=${t}&o=8&p=8&l=as1&asins=${asins}&fc1=${fc1}&IS2=${IS2}&lt1=${lt1}&m=amazon&lc1=${lc1}&bc1=${bc1}&bg1=${bg1}&f=ifr" style="width:120px;height:240px;" scrolling="no" marginwidth="0" marginheight="0" frameborder="0"></iframe>',
    ),
    'JP' => array(
        'domain'             => 'Amazon.co.jp',
        'baseUri'            => 'https://webservices.amazon.co.jp/onca/xml',
        'linkTemplate'       => '<iframe src="https://rcm-jp.amazon.co.jp/e/cm?t=${t}&o=9&p=8&l=as1&asins=${asins}&fc1=${fc1}&IS2=${IS2}&lt1=${lt1}&m=amazon&lc1=${lc1}&bc1=${bc1}&bg1=${bg1}&f=ifr" style="width:120px;height:240px;" scrolling="no" marginwidth="0" marginheight="0" frameborder="0"></iframe>',
    ),
    'CA' => array(
        'domain'             => 'Amazon.ca',
        'baseUri'            => 'https://webservices.amazon.ca/onca/xml',
        'linkTemplate'       => '<iframe src="https://rcm-ca.amazon.ca/e/cm?t=${t}&o=15&p=8&l=as1&asins=${asins}&fc1=${fc1}&IS2=${IS2}&lt1=${lt1}&m=amazon&lc1=${lc1}&bc1=${bc1}&bg1=${bg1}&f=ifr" style="width:120px;height:240px;" scrolling="no" marginwidth="0" marginheight="0" frameborder="0"></iframe>',
    ),
    'CN' => array(
        'domain'             => 'Amazon.cn',
        'baseUri'            => 'https://webservices.amazon.cn/onca/xml',
        'linkTemplate'       => '<iframe src="https://rcm-cn.amazon.cn/e/cm?t=${t}&o=28&p=8&l=as1&asins=${asins}&fc1=${fc1}&IS2=${IS2}&lt1=${lt1}&m=amazon&lc1=${lc1}&bc1=${bc1}&bg1=${bg1}&f=ifr" style="width:120px;height:240px;" scrolling="no" marginwidth="0" marginheight="0" frameborder="0"></iframe>',
    ),
    'IT' => array(
        'domain'             => 'Amazon.it',
        'baseUri'            => 'https://webservices.amazon.it/onca/xml',
        'linkTemplate'       => '<iframe src="https://rcm-it.amazon.it/e/cm?t=${t}&o=29&p=8&l=as1&asins=${asins}&fc1=${fc1}&IS2=${IS2}&lt1=${lt1}&m=amazon&lc1=${lc1}&bc1=${bc1}&bg1=${bg1}&f=ifr" style="width:120px;height:240px;" scrolling="no" marginwidth="0" marginheight="0" frameborder="0"></iframe>',
    ),
    'ES' => array(
        'domain'             => 'Amazon.es',
        'baseUri'            => 'https://webservices.amazon.es/onca/xml',
        'linkTemplate'       => '<iframe src="https://rcm-es.amazon.es/e/cm?t=${t}&o=30&p=8&l=as1&asins=${asins}&fc1=${fc1}&IS2=${IS2}&lt1=${lt1}&m=amazon&lc1=${lc1}&bc1=${bc1}&bg1=${bg1}&f=ifr" style="width:120px;height:240px;" scrolling="no" marginwidth="0" marginheight="0" frameborder="0"></iframe>',
    ),
);

$locale = array_key_exists ( User_Locale, $locales ) ? $locales[User_Locale] : $locales['US'];
$baseurl = $locale['baseUri'];

// Request parameters
$params = array();
$params["Service"]        = "AWSECommerceService";
$params["AWSAccessKeyId"] = Access_Key_ID;
$params["Version"]        = "2013-08-01";
$params["Operation"]      = "ItemLookup";
$params["ItemId"]         = $ItemId;
$params["AssociateTag"]   = Associate_Tag;
$params["ResponseGroup"]  = "ItemAttributes,Offers,Images";
$params["Timestamp"]      = gmdate("Y-m-d\TH:i:s\Z");

// Alphabetical order
ksort($params);

// Concatenate parameters
$url_param = "";
foreach ($params as $k => $v) {
    $url_param .= '&' . $k . '=' . rawurlencode($v);
}
$url_param = substr($url_param, 1);

// Signature
$parsed_url = parse_url($baseurl);
$sigbase = "GET\n" . $parsed_url['host'] . "\n" . $parsed_url['path'] . "\n" . $url_param;
$signature = base64_encode(hash_hmac('sha256', $sigbase, Secret_Access_Key, true));
$signature = rawurlencode($signature);

// Create request
$request = $baseurl . "?" . $url_param . "&Signature=" . $signature;

// echo '<a href="'.$request.'">link</a>';


amazon-products.php
<?php
/**
 * Short description for amazon-products.php
 *
 * @package amazon-products
 * @author Kazuya Kanatani
 * @version 0.1
 * @copyright (C) 2017 kinformation<kanatani.social@gmail.com>
 * @license MIT
 */

namespace Grav\Plugin;

use Grav\Common\Plugin;
// use Grav\Common\Data\Data;
// use Grav\Common\Page\Page;
// use Grav\Common\Twig\Twig;
// use Grav\Plugin\Youtube\Twig\YoutubeTwigExtension;
// use RocketTheme\Toolbox\Event\Event;

class AmazonProductsPlugin extends Plugin
{
    /**
     * @return array
     */
    public static function getSubscribedEvents()
    {
        return [
            'onShortcodeHandlers' => ['onShortcodeHandlers', 0],
            'onTwigTemplatePaths' => ['onTwigTemplatePaths', 0],
            'onTwigSiteVariables' => ['onTwigSiteVariables', 0],
        ];
    }

    /**
     * Add style and script to page.
     */
    public function onTwigSiteVariables()
    {
        if (!$this->isAdmin()) {
            $this->grav['assets']->addCss('plugin://amazon-products/assets/css/amazon-products.css');
        }
        if ($this->isAdmin() && $this->config->get('plugins.amazon-products.editor_button')) {
            $this->grav['assets']->add('plugin://amazon-products/admin/editor-button/js/button.js');
        }
    }

    /**
     * Add current directory to twig lookup paths.
     */
    public function onTwigTemplatePaths()
    {
        $this->grav['twig']->twig_paths[] = __DIR__ . '/templates';
    }

    /**
     * Initialize configuration
     */
    public function onShortcodeHandlers()
    {
        $this->grav['shortcode']->registerAllShortcodes(__DIR__.'/shortcodes');
    }
}


shortcodes/AmazonShortcode.php
<?php
/**
 * Short description for AmazonShortcode.php
 *
 * @package amazon
 * @author Kazuya Kanatani
 * @version 0.1
 * @copyright (C) 2017 kinformation<kanatani.social@gmail.com>
 * @license MIT
 */

namespace Grav\Plugin\Shortcodes;

use Thunder\Shortcode\Shortcode\ShortcodeInterface;

class AmazonShortcode extends Shortcode
{
    public function init()
    {
        $this->shortcode->getHandlers()->add('amazon', function (ShortcodeInterface $sc) {

            $current_date = date("Y/m/d H:i");

            $asin = $sc->getParameter('asin');
            $item = $this->ItemLookup($asin);
            if (empty($item)) {
                return '';
            }

            $item_title = $item->ItemAttributes->Title;
            $item_image  = $item->MediumImage->URL;
            $item_url = $item->DetailPageURL;
            $item_price = $item->Offers->TotalOffers > 0
                ? $item->Offers->Offer->OfferListing->Price->FormattedPrice
                : $item->OfferSummary->LowestNewPrice->FormattedPrice;

            $item_productgroup = $item->ItemAttributes->ProductGroup;
            switch ($item_productgroup) {
                case 'Book':
                    $item_author = $item->ItemAttributes->Author;
                    $item_publicationdate = $item->ItemAttributes->PublicationDate;
                    $item_publisher = $item->ItemAttributes->Publisher;
                    $output = $this->twig->processTemplate('partials/amazon_book.html.twig', [
                        'amazon_date' => $current_date,
                        'amazon_title' => $item_title,
                        'amazon_url' => $item_url,
                        'amazon_price' => $item_price,
                        'amazon_image' => $item_image,
                        'amazon_author' => $item_author,
                        'amazon_publicationDate' => $item_publicationdate,
                        'amazon_publisher' => $item_publisher,
                    ]);
                    break;

                case 'Music':
                case 'Digital Music Album':
                    $item_artist = $item->ItemAttributes->Artist;
                    if (empty($item_artist))
                        $item_artist = $item->ItemAttributes->Creator;
                    $item_label =  $item->ItemAttributes->Label;
                    $item_releaseDate = $item->ItemAttributes->ReleaseDate;
                    $item_runningTime =  $item->ItemAttributes->RunningTime;
                    $item_timeUnit =  $item->ItemAttributes->RunningTime['Units'];
                    $output = $this->twig->processTemplate('partials/amazon_music.html.twig', [
                        'amazon_date' => $current_date,
                        'amazon_title' => $item_title,
                        'amazon_url' => $item_url,
                        'amazon_price' => $item_price,
                        'amazon_image' => $item_image,
                        'amazon_artist' => $item_artist,
                        'amazon_label' => $item_label,
                        'amazon_releaseDate' => $item_releaseDate,
                        'amazon_runningTime' => $item_runningTime,
                        'amazon_timeUnit' => $item_timeUnit,
                    ]);
                    break;

                case 'DVD':
                    $item_director = $item->ItemAttributes->Director;
                    $item_actor = $item->ItemAttributes->Actor;
                    $item_label = $item->ItemAttributes->Label;
                    $item_releaseDate = $item->ItemAttributes->ReleaseDate;
                    $item_runningTime =  $item->ItemAttributes->RunningTime;
                    $item_timeUnit =  $item->ItemAttributes->RunningTime['Units'];
                    $output = $this->twig->processTemplate('partials/amazon_video.html.twig', [
                        'amazon_date' => $current_date,
                        'amazon_title' => $item_title,
                        'amazon_url' => $item_url,
                        'amazon_price' => $item_price,
                        'amazon_image' => $item_image,
                        'amazon_director' => $item_director,
                        'amazon_actor' => $item_actor,
                        'amazon_label' => $item_label,
                        'amazon_releaseDate' => $item_releaseDate,
                        'amazon_runningTime' => $item_runningTime,
                        'amazon_timeUnit' => $item_timeUnit,
                    ]);
                    break;

                default:
                    $item_brand = $item->ItemAttributes->Brand;
                    $output = $this->twig->processTemplate('partials/amazon_default.html.twig', [
                        'amazon_date' => $current_date,
                        'amazon_title' => $item_title,
                        'amazon_url' => $item_url,
                        'amazon_price' => $item_price,
                        'amazon_image' => $item_image,
                        'amazon_brand' => $item_brand,
                    ]);
            }

            return $output;
        });
    }

    private function ItemLookup($ItemId)
    {
        if (!defined('User_Locale'))
            define("User_Locale", strtoupper($this->config->get('plugins.amazon.locale')));
        if (!defined('Associate_Tag')) {
            $atag = trim($this->config->get('plugins.amazon.keys.associateTag'));
            !empty($atag)
                ? define("Associate_Tag", $atag)
                : define("Associate_Tag", 'XXXXX');
        }
        if (!defined('Access_Key_ID'))
            define("Access_Key_ID", trim($this->config->get('plugins.amazon.keys.accessKeyId')));
        if (!defined('Secret_Access_Key'))
            define("Secret_Access_Key", trim($this->config->get('plugins.amazon.keys.secretAccessKey')));
        include("plugin://amazon/amazon_request.php");

        $amazon_xml = '';
        $ret = '';
        for ($i = 0; $i < 2; $i++) {
            $context = stream_context_create(array(
                'http' => array('ignore_errors' => true)
            ));
            $response = file_get_contents($request, false, $context);

            $pos = strpos($http_response_header[0], '200');
            if ($pos === false) {
                continue;
            }
            $amazon_xml = simplexml_load_string($response);
            $ret = $amazon_xml === false ? '' : $amazon_xml->Items->Item;
            break;
        }
        return $ret;
    }
}


admin/editor-button/js/button.js
(function($) {
    $(function() {
        $('body').on('grav-editor-ready', function() {
            var Instance = Grav.default.Forms.Fields.EditorField.Instance;
            Instance.addButton({
                youtube: {
                    identifier: 'amazon-products',
                    title: 'Amazon Products',
                    label: '<i class="fa fa-fw fa-amazon"></i>',
                    modes: ['gfm', 'markdown'],
                    action: function(_ref) {
                        var codemirror = _ref.codemirror,
                            button = _ref.button,
                            textarea = _ref.textarea;
                        button.on('click.editor.amazon-products', function() {
                            var input = prompt("ASIN/ISBN-10/Amazon Products URL");
                            asin = checkASIN(input);
                            if (asin) {
                                var text = '[amazon asin=' + asin + '][/amazon]';

                                //Add text to the editor
                                var pos = codemirror.getDoc().getCursor(true);
                                var posend = codemirror.getDoc().getCursor(false);

                                for (var i = pos.line; i < (posend.line + 1); i++) {
                                    codemirror.replaceRange(text + codemirror.getLine(i), { line: i, ch: 0 }, { line: i, ch: codemirror.getLine(i).length });
                                }

                                codemirror.setCursor({ line: posend.line, ch: codemirror.getLine(posend.line).length });
                                codemirror.focus();
                            }
                        });
                    }
                }
            });

            function checkASIN(input) {
                // Only ASIN/ISBN
                if (input.match(/^[a-zA-Z0-9]+$/)) {
                    return input.match(/^[a-zA-Z0-9]+$/);
                }

                // Extract ASIN by raw amazonURL
                if (input.match(/\/dp\/([a-zA-Z0-9]+)\//)) {
                    return input.match(/\/dp\/([a-zA-Z0-9]+)\//)[1];
                }
                if (input.match(/\/gp\/product\/([a-zA-Z0-9]+)\//)) {
                    return input.match(/\/gp\/product\/([a-zA-Z0-9]+)\//)[1];
                }

                return input;
            };
        });
    });
})(jQuery);

shortcodes/AmazonProductsShortcode.php
<?php
/**
 * Short description for AmazonProductsShortcode.php
 *
 * @package amazon-products
 * @author Kazuya Kanatani
 * @version 0.1
 * @copyright (C) 2017 kinformation<kanatani.social@gmail.com>
 * @license MIT
 */

namespace Grav\Plugin\Shortcodes;

use Thunder\Shortcode\Shortcode\ShortcodeInterface;

class AmazonProductsShortcode extends Shortcode
{
    public function init()
    {
        $this->shortcode->getHandlers()->add('amazon', function (ShortcodeInterface $sc) {

            $current_date = date("Y/m/d H:i");

            $asin = $sc->getParameter('asin');
            $asin = $this->extract_asin($asin);
            $item = $this->lookup_item($asin);
            if (empty($item)) {
                return '';
            }

            $item_title = $item->ItemAttributes->Title;
            $item_image  = $item->MediumImage->URL;
            $item_url = $item->DetailPageURL;
            $item_price = @$item->Offers->Offer->OfferListing->Price->FormattedPrice
                       ?: @$item_price = $item->OfferSummary->LowestNewPrice->FormattedPrice
                       ?: '';
            $item_productgroup = $item->ItemAttributes->ProductGroup;
            switch ($item_productgroup) {
                case 'Book':
                case 'eBooks':
                    $item_author = $item->ItemAttributes->Author;
                    $item_publicationdate = $item->ItemAttributes->PublicationDate;
                    $item_publisher = $item->ItemAttributes->Publisher;
                    $output = $this->twig->processTemplate('partials/amazon_book.html.twig', [
                        'amazon_date' => $current_date,
                        'amazon_title' => $item_title,
                        'amazon_url' => $item_url,
                        'amazon_price' => $item_price,
                        'amazon_image' => $item_image,
                        'amazon_author' => $item_author,
                        'amazon_publicationDate' => $item_publicationdate,
                        'amazon_publisher' => $item_publisher,
                    ]);
                    break;

                case 'Music':
                case 'Digital Music Album':
                case 'Digital Music Track':
                    $item_artist = $item->ItemAttributes->Artist;
                    if (empty($item_artist)) {
                        $item_artist = $item->ItemAttributes->Creator;
                    }
                    $item_label =  $item->ItemAttributes->Label;
                    $item_releaseDate = $item->ItemAttributes->ReleaseDate;
                    $item_runningTime =  $item->ItemAttributes->RunningTime;
                    $item_timeUnit =  $item->ItemAttributes->RunningTime['Units'];
                    $output = $this->twig->processTemplate('partials/amazon_music.html.twig', [
                        'amazon_date' => $current_date,
                        'amazon_title' => $item_title,
                        'amazon_url' => $item_url,
                        'amazon_price' => $item_price,
                        'amazon_image' => $item_image,
                        'amazon_artist' => $item_artist,
                        'amazon_label' => $item_label,
                        'amazon_releaseDate' => $item_releaseDate,
                        'amazon_runningTime' => $item_runningTime,
                        'amazon_timeUnit' => $item_timeUnit,
                    ]);
                    break;

                case 'DVD':
                case 'Movie':
                    $item_director = $item->ItemAttributes->Director;
                    $item_actor = $item->ItemAttributes->Actor;
                    $item_label = $item->ItemAttributes->Label;
                    $item_releaseDate = $item->ItemAttributes->ReleaseDate;
                    $item_runningTime =  $item->ItemAttributes->RunningTime;
                    $item_timeUnit =  $item->ItemAttributes->RunningTime['Units'];
                    $output = $this->twig->processTemplate('partials/amazon_video.html.twig', [
                        'amazon_date' => $current_date,
                        'amazon_title' => $item_title,
                        'amazon_url' => $item_url,
                        'amazon_price' => $item_price,
                        'amazon_image' => $item_image,
                        'amazon_director' => $item_director,
                        'amazon_actor' => $item_actor,
                        'amazon_label' => $item_label,
                        'amazon_releaseDate' => $item_releaseDate,
                        'amazon_runningTime' => $item_runningTime,
                        'amazon_timeUnit' => $item_timeUnit,
                    ]);
                    break;

                default:
                    $item_brand = $item->ItemAttributes->Brand;
                    $output = $this->twig->processTemplate('partials/amazon_default.html.twig', [
                        'amazon_date' => $current_date,
                        'amazon_title' => $item_title,
                        'amazon_url' => $item_url,
                        'amazon_price' => $item_price,
                        'amazon_image' => $item_image,
                        'amazon_brand' => $item_brand,
                    ]);
            }

            return $output;
        });
    }

    private function lookup_item($ItemId)
    {
        if (!defined('User_Locale')) {
            define("User_Locale", strtoupper($this->config->get('plugins.amazon-products.locale')));
        }
        if (!defined('Associate_Tag')) {
            $atag = trim($this->config->get('plugins.amazon-products.keys.associateTag'));
            !empty($atag)
                ? define("Associate_Tag", $atag)
                : define("Associate_Tag", 'XXXXX');
        }
        if (!defined('Access_Key_ID')) {
            define("Access_Key_ID", trim($this->config->get('plugins.amazon-products.keys.accessKeyId')));
        }
        if (!defined('Secret_Access_Key')) {
            define("Secret_Access_Key", trim($this->config->get('plugins.amazon-products.keys.secretAccessKey')));
        }
        include("plugin://amazon-products/amazon_request.php");

        $amazon_xml = '';
        $ret = '';
        for ($i = 0; $i < 2; $i++) {
            $context = stream_context_create(array(
                'http' => array('ignore_errors' => true)
            ));
            $response = file_get_contents($request, false, $context);

            $pos = strpos($http_response_header[0], '200');
            if ($pos === false) {
                continue;
            }
            $amazon_xml = simplexml_load_string($response);
            $ret = $amazon_xml === false ? '' : $amazon_xml->Items->Item;
            break;
        }
        return $ret;
    }

    private function extract_asin($asin)
    {
        if (empty($asin)) {
            return '';
        }

        // Only ASIN/ISBN
        if (preg_match('/^[a-zA-Z0-9]+$/', $asin) === 1) {
            return $asin;
        }

        // Extract ASIN by raw amazonURL
        if (preg_match('#/dp/([a-zA-Z0-9]+)/#', $asin, $matches) === 1) {
            return $matches[1];
        }
        if (preg_match('#/gp/product/([a-zA-Z0-9]+)/#', $asin, $matches) === 1) {
            return $matches[1];
        }

        return '';
    }
}
